Mathlib/GroupTheory/Congruence/Hom.lean

```diff
@@ -33,30 +33,134 @@ variable {M}
 
 namespace Con
 
-section MulOneClass
+section Mul
+variable {F} [Mul M] [Mul N] [Mul P] [FunLike F M N] [MulHomClass F M N]
 
-variable [MulOneClass M] [MulOneClass N] [MulOneClass P] {c : Con M}
 
-/-- The kernel of a monoid homomorphism as a congruence relation. -/
-@[to_additive "The kernel of an `AddMonoid` homomorphism as an additive congruence relation."]
-def ker (f : M →* P) : Con M :=
-  mulKer f (map_mul f)
+/-- The natural homomorphism from a magma to its quotient by a congruence relation. -/
+@[to_additive (attr := simps)"The natural homomorphism from an additive magma to its quotient by an
+additive congruence relation."]
+def mkMulHom (c : Con M) : MulHom M c.Quotient where
+  toFun := (↑)
+  map_mul' _ _ := rfl
+
+
+/-- The kernel of a multiplicative homomorphism as a congruence relation. -/
+@[to_additive "The kernel of an additive homomorphism as an additive congruence relation."]
+def ker (f : F) : Con M where
+  toSetoid := Setoid.ker f
+  mul' h1 h2 := by
+    dsimp [Setoid.ker, onFun] at *
+    rw [map_mul, h1, h2, map_mul]
+
+@[to_additive (attr := norm_cast)]
```

Comment:
Should it be `simp` as well?

Reply:
Arguably the LHS should be simp-normal, since then you only need to write one `ker_comp` simp lemma for `MulHom.comp` and not one for every `comp`. But I don't want to open that can of worms here.